design-proposals/klip-8-interactive-queries.md

```diff
@@ -0,0 +1,143 @@
+
+Here we roughly subdivide the spectrum of interactive query complexity into five progressive tiers, each tier building on the one before it. Each of these tiers can be released separately, although it may make sense to combine some of them.
+
+*Please note that some of the example queries below use mock functions for illustrative purposes.*
+
+### Tier 0 - Simple key-based lookups
+
+Support for key-based lookups is the bare minimum starting point for interactive queries that will still add meaningful utility. This essentially amounts to a cosmetic SQL layer above RocksDB’s key-value interface. Key-based lookups will allow users to look up a single row, given a row key:
+
+* WHERE clause must be strictly of the form: `WHERE ROWKEY = <rowkey>`
+* Each query can therefore only target a single shard.
+* Arbitrary expressions should be allowed.
+* Examples:
+  * `SELECT x, y, z FROM table WHERE ROWKEY = ‘rowkey’`
+  * `SELECT x + y FROM table WHERE ROWKEY = ‘rowkey’`
+
+### Tier 1 - Key-based range queries
+
+Key-based range queries are a minimal incremental step forward from single-shard queries towards multi-shard queries. Range queries may target multiple shards, but restricting them to be key based will allow us to leverage RocksDB’s native range scan support instead of doing extraneous row scanning:
+
+* Queries with a predicate of the form WHERE ROWKEY > lower_bound AND ROWKEY < upper_bound should be allowed.
+* While not technically a range query, queries with a WHERE ROWKEY IN … predicate should be allowed. These are multi-key lookup queries.
+* Examples:
+  * `SELECT … FROM table WHERE ROWKEY >= 0 AND ROWKEY < 10000`
+  * `SELECT … FROM table WHERE ROWKEY < 1000`
+  * `SELECT … FROM table WHERE ROWKEY IN (‘k0’, ‘k1’, ‘k2’)`
+
+### Tier 2 - Generic multi-shard queries
+
+While range queries may target multiple shards, expanding to generic multi-shard queries will add significant flexibility to the supported interactive query forms. Multi-shard queries should be able to target any selection of RocksDB rows via comparisons performed aginst any columns:
+
+* Arbitrary WHERE clauses should be allowed.
+* Examples:
+  * `SELECT … FROM table WHERE x = ‘x’ AND y = ‘y’`
+  * `SELECT … FROM table WHERE str LIKE ‘%substr%’`
+  * `SELECT … FROM table WHERE x < y`
+
+### Tier 3 - Aggregates
+
+Aggregations are central to extracting value from datasets.
+
+While materialized views in KSQL may already be defined using aggregates, it will still be very useful to perform further aggregations over them via interactive queries. For example, users may want to maintain a materialized view summarized at the minute level but then summarize it further down to the hour level to power a frontend visualization that dynamically supplies a date range:
```

Comment:
I'm not convinced such a use-case would use interactive aggregates.

If I was looking to implement such a use-case I would build the minute level view and build an hourly level one from the minute level one.  The window retention could be different for the two views, e.g. the minute one could have a retention of a few hours and the hourly one a few days.

This would provide a much better performing system than having the database constantly computing aggregates over mostly historic data.

There may be use-cases out there that justify interactive aggregates, but I'm struggling to think of any.  I see this functionality as a long way down the priority list, simply because the same can be achieved with persistent aggregates as it stands today.

Reply:
You would definitely be able to aggregate at different levels using the method you described. But the downside there is that if you created a minute-level materialized view, and then much later you wanted hour-level aggregate data, you wouldn't get the history that the minute-level materialized view has. And consuming from `earliest` isn't really an option here either since the hour-level materialized view isn't based on the original input topic/stream.

Also, changing the timestamp granularity is just one example of why you'd want to interactively aggregate an existing materialized view. But the underlying concept is much more generally applicable than that: changing the aggregation grouping.

For example, consider this materialized view:

```sql
CREATE TABLE agg AS SELECT city, region, country, count(*)
  FROM s
GROUP BY city, region, country
```
And every so often you want to see aggregate values broken down only by `region`:
```sql
SELECT region, sum(count) FROM agg GROUP BY region;
```
You wouldn't always necessarily want to create a materialized view for every possible aggregation level you foresee being required, so I think this capability adds a lot of flexibility on top of the rigidity of predeclared materialized views.

And to make this *really* useful, we can potentially eventually add functionality for [combining](http://docs.pipelinedb.com/aggregates.html#combine) aggregate values with no loss of information. Obviously this is trivial with scalar aggregates such as `count`, you just `sum` the count. But for more complex aggregates more advanced infrastructure would be needed.

Reply:
I guess this comes down to different use-cases, e.g. production SLA based queries to service some web tier or microsevice, vs ad-hoc queries.

I'd certainly be focusing our attention on the former, as this is likely our biggest strength and use-case. But, yes, in time we could support the ad-hoc stuff.